### PR TITLE
Add App Engine directory field to flex deploy dialog

### LINKED ISSUE #1823: Retrieve App Engine directory from flex deploy preferences (set via flex deploy dialog) when executing flex deploy command
When deploying flex projects, we assume the App Engine directory (where `app.yaml` is located) is always `src/main/appengine`, but this should come from the flex deploy preference (`FlexDeployPreferences.getAppEngineDirectory`) which will be set by the deploy dialog. The dialog also needs to be extended to have a path input field for this directory.

### PR DESCRIPTION
**Pull request by chanseokoh**
Fixes #1823 #1685.

- If the value in the field is a relative path (i.e., does not start with '/'), the path is considered relative to the project root.
- If a user chooses a directory from the directory choose dialog, the field will be set to a path that wil be  relative to the project root (e.g., `src/main/appengine` instead of `/usr/home/alice/workspace/my-project/src/main/appengine`).
- Validation on the field is triggered only on the deploy dialog (i.e., not triggered on the pref page).

deploy to app engine flexible _013

- I think using '/' everywhere and `toString()` instead of `toOSString()` will work on Windows too, but it'd be nice to do some actual testing.

### CODE REVIEW

**Comment by chanseokoh** [comment]
Right, "App Engine directory" is too generic to the point of being anything related to App Engine. "App Engine deploy configuration directory"? Or, "Directory containing app.yaml"?

**Comment by elharo** [comment]
Why are we selecting the directory instead of the app.yaml file? Perhaps we can make the chooser find the app.yaml file, and then the field label becomes app.yaml.

**Comment by elharo** [comment]
Your screenshot is `src/main/appengin`, not `src/main/appengine`. Is this #1830 again? If so, we may need to handle that.

**Comment by chanseokoh** [comment]
The clipped `appengin` (no `e`) was to show the validation error. It is working fine.

**Comment by chanseokoh** [comment]
Ready for review. It's been known that Kokoro builds fail quite frequently on a test that is independent of this PR.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanelTest.java [hunk 1]
```diff
@@ -17,30 +17,178 @@
...
+    appYamlField.setText("non/existing/app.yaml");
+    assertFalse(hasValidationError(panel));
+  }
+
+  @Test
+  public void testAppYamlPathValidation_absoluatePathWorks() throws InterruptedException {
+    FlexDeployPreferencesPanel panel = createPanel(false /* requireValues */);
+    Text appYamlField = findAppYamlField(panel);
+
+    IPath absoluatePath = project.getLocation().append("src/main/appengine/app.yaml");
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanelTest.java]: sp: absolutePath

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanelTest.java [hunk 2]
```diff
@@ -17,30 +17,178 @@
...
+      throws InterruptedException {
+    FlexDeployPreferencesPanel panel = createPanel(false /* requireValues */);
+
+    Text appYamlField = findAppYamlField(panel);
+    appYamlField.setText("non/existing/app.yaml");
+    assertFalse(hasValidationError(panel));
+  }
+
+  @Test
+  public void testAppYamlPathValidation_absoluatePathWorks() throws InterruptedException {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanelTest.java]: sp: absolutePath

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanelTest.java [hunk 3]
```diff
@@ -17,30 +17,178 @@
...
+      }
+    }
+    return false;
+  }
+
+  private static Text findAppYamlField(Composite panel) {
+    return (Text) findControl(panel, new Predicate<Control>() {
+      @Override
+      public boolean apply(Control control) {
+        return control instanceof Text && APP_YAML_FIELD_TOOLTIP.equals(control.getToolTipText());
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanelTest.java]: I'm surprised there's no ID on the control, though this is a very common omission in UI toolkits. This works well enough for a test, I suppose.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/AppYamlPathValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    basePath = new Path(tempFolder.newFolder().toString());
+    assertTrue(basePath.isAbsolute());
+  }
+
+  @Test
+  public void testContructor_nonAbsoluteBasePath() {
+    try {
+      new AppYamlPathValidator(new Path("non/absolute/base/path"));
+      fail();
+    } catch (IllegalArgumentException ex) {}
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/AppYamlPathValidatorTest.java]: This feels like the wrong exception type here; more below

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/AppYamlPathValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  public void testValidate_relativePathAndInvalidFileName() {
+    IStatus result = new AppYamlPathValidator(basePath).validate("relative/path/my-app.yaml");
+    assertEquals(IStatus.ERROR, result.getSeverity());
+    assertEquals("File name is not app.yaml: "
+        + new Path(basePath + "/relative/path/my-app.yaml").toOSString(),
+        result.getMessage());
+  }
+
+  @Test
+  public void testValidate_absolautePathInvalidFileName() {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/AppYamlPathValidatorTest.java]: absolute

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/AppYamlPathValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+  @Test
+  public void testValidate_relativePathAndNoAppYaml() {
+    IStatus result = new AppYamlPathValidator(basePath).validate("relative/path/app.yaml");
+    assertEquals(IStatus.ERROR, result.getSeverity());
+    assertEquals("app.yaml does not exist.", result.getMessage());
+  }
+
+  @Test
+  public void testValidate_absoluatePathAndNoAppYaml() {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/AppYamlPathValidatorTest.java]: absolute

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/AppYamlPathValidatorTest.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    createAppYamlAsDirectory(basePath);
+
+    IStatus result = new AppYamlPathValidator(basePath).validate("app.yaml");
+    assertEquals(IStatus.ERROR, result.getSeverity());
+    assertEquals("Not a file: " + new Path(basePath + "/app.yaml").toOSString(),
+        result.getMessage());
+  }
+
+  @Test
+  public void testValidate_absolautePathNotFile() {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/AppYamlPathValidatorTest.java]: absolute

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/AppEngineDeployPreferencesPanel.java [hunk 8]
```diff
@@ -112,13 +112,14 @@
 
   private final Image refreshIcon = SharedImages.REFRESH_IMAGE_DESCRIPTOR.createImage(getDisplay());
 
+  protected final IProject project;
   @VisibleForTesting
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/AppEngineDeployPreferencesPanel.java]: probably no longer visible for testing if it's protected

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/AppYamlPathValidator.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+import com.google.cloud.tools.eclipse.appengine.deploy.ui.Messages;
+import com.google.common.base.Preconditions;
+import java.io.File;
+import org.eclipse.core.databinding.validation.IValidator;
+import org.eclipse.core.databinding.validation.ValidationStatus;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IStatus;
+
+public class AppYamlPathValidator implements IValidator {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/AppYamlPathValidator.java]: javadoc please

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties [hunk 10]
```diff
@@ -10,7 +10,9 @@ deploy.preferences.dialog.title.standard=Deploy to App Engine Standard
 deploy.preferences.dialog.title.flexible=Deploy to App Engine Flexible
 deploy.preferences.dialog.title.withProject=Deployment Parameters for "{0}"
 deploy.preferences.dialog.label.selectAccount=Account:
+deploy.preferences.dialog.label.app.yaml=app.yaml:
 deploy.preferences.dialog.accountSelector.login=<Add a new account...>
+deploy.preferences.dialog.browser.dir=Browse
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties]: still ".dir"?

Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties]: Oops. It's also `browser`.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanel.java [hunk 11]
```diff
@@ -45,6 +59,40 @@ protected void createCenterArea() {
...
   }
 
+  private void createAppYamlSection() {
+    // Part 1. create UI widgets
+    Label label = new Label(this, SWT.LEAD);
+    label.setText(Messages.getString("deploy.preferences.dialog.label.app.yaml"));
+    label.setToolTipText(Messages.getString("tooltip.app.yaml"));
+
+    Composite secondColumn = new Composite(this, SWT.NONE);
+    final Text appYamlField = new Text(secondColumn, SWT.SINGLE | SWT.BORDER);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanel.java]: I don't think this needs to be final.

File: plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/.classpath [hunk 12]
```diff
@@ -1,10 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
-	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.7">
-		<attributes>
-			<attribute name="maven.pomderived" value="true"/>
-		</attributes>
-	</classpathentry>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.7"/>
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/.classpath]: Is this change really part of this PR?

**Comment by chanseokoh** [comment]
This should be good in itself. PTAL. (Ignore the Kokoro failure. It's too flaky and has nothing to do with this PR: 
 https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1896)

**Pull request merged**